### DISCUSSION TITLE
How to set Sentry Config in asp.net core IMiddleware

### DISCUSSION THREAD

**Discussion by anwarjaved** [getsentry/sentry-dotnet#1730] [Q&A] [answered]
I want to use same code on multiple domain and want to set different dsn according to domain, so I need to configure sentry dsn for it, is setting SENTRY_DSN and SENTRY_ENVIRONMENT will work for me if I set it in middleware

**Comment by mattjohnsonpint** [comment]
Hi @anwarjaved, thanks for your question.  We don't allow the DSN to be changed while an application is running.  It is expected that a single instance of the Sentry SDK will be used, and the same DSN will be used the entire time.

In an ASP.NET Core application, you can set the DSN in your `appsettings.json` file.  Just like any other setting you may have in that file, it can be configured per *environment* you might deploy to (production, staging, development, etc.), but not per domain.

**Reply by anwarjaved** [reply]
Hi @mattjohnsonpint 

Thanks for your reply, I have an api which is behaving according to the domain on which it is hosted, so I need to call UseSentry according to requested Domain, is that possible?

**Reply by mattjohnsonpint** [reply]
If you mean that one instance of ASP.NET Core is serving multiple domains, then no - unfortunately that is not currently possible because the DSN is a global setting, not something that can be changed per incoming HTTP request.

If you just mean that you deploy the same code to multiple different web servers and each one serves a different domain, then you could change the DSN in your app settings during deployment.

But I presume you mean the first scenario.

**Reply by mattjohnsonpint** [reply]
I'll explain, so you understand why we can't currently support multiple DSNs.

The Sentry SDK doesn't send data to Sentry _immediately_ when an event occurs.  Instead, it creates "envelopes" from events and buffers them into a queue to be sent later.  There's a background thread that reads from the queue to send the envelopes to Sentry.  This allows your app to continue doing its work instead of pausing while waiting to talk to Sentry.

The DSN is read from the options when the envelope is actually sent to Sentry. It's placed into the `X-Sentry-Auth` HTTP header, and it controls the actual destination URL that the data is sent to.

https://github.com/getsentry/sentry-dotnet/blob/99ba1b78b49692071dc81782abbe7dce370eb341/src/Sentry/Http/HttpTransportBase.cs#L183-L196

If we were to support multiple different DSNs in the same instance, then we would have to embed the DSN (or some other distinguishing information) into the envelope itself.  That would break a few things, and would have storage consequences (especially if caching to disk is enabled).  It would also need to be removed from the envelope before being sent along to Sentry, so would have some performance overhead.

In short, while it's technically possible - it's not trivial and it's not a scenario we plan to support.

It also doesn't align with how most users use Sentry.  Since the data contains exceptions and performance data that is about the application code, it's generally thought that the application developer would want all their data to go to a single place so it can be reviewed and analyzed together, even for a multi-tenant web application.

If ultimately you're just trying to distinguish errors that are occurring on a specific domain, you can always filter by url in the Sentry dashboard.  For example:  `is:unresolved url:*//somedomain.com/*`

**Reply by anwarjaved** [reply]
What if I call SentrySDk config in middleware `sing (SentrySdk.Init(o =>
            {
                // Tells which project in Sentry to send events to:
                o.Dsn = configuration["Sentry:Dsn"];
                // When configuring for the first time, to see what the SDK is doing:
                o.Debug = true;
                // Set traces_sample_rate to 1.0 to capture 100% of transactions for performance monitoring.
                // We recommend adjusting this value in production.
                o.TracesSampleRate = 1.0;
            }))
            {
                // App code goes here - Disposing will flush events out
            }`

and capture exception like this

`try
            {
               await next(context);
            }
            catch (Exception ex)
            {
               SentrySdk.CaptureException(ex);
            }`
Will it work?
I am using .Net 6 C# Web Api,

**Reply by mattjohnsonpint** [reply]
No, you should not do that.

You should call UseSentry in the host builder.  It adds our own middleware, which initializes once for the entire application. This is the only usage we support or recommend for aspnet core. 

If you tried to initialize Sentry in your own middleware per request, the overhead would be horrible and you would likely have one request stepping on another.

**Reply by anwarjaved** [reply]
Can we use SENTRY_DSN environment variable here?

**Reply by mattjohnsonpint** [reply]
You can use the `SENTRY_DSN` environment variable, yes.   However, it's just another way to pass the DSN to the application and is read once at startup.  You can't change it on the fly to switch DSNs.  I appreciate the out of the box thinking though.

Just to be perfectly clear - We do not support using multiple DSN in the same process.  This was a fundamental design decision, and we believe it is the correct one.

**Reply by anwarjaved** [reply]
@mattjohnsonpint ok got it 
but can we pass custom tags according to the domain? is that possible or we still have limitations there too.

**Reply by mattjohnsonpint** [reply]
Absolutely - you can set tags any way you like.  They are set on the current scope, so they can contain information that's per-request without any issues.  Example here:
https://docs.sentry.io/platforms/dotnet/guides/aspnetcore/enriching-events/tags/

Keep in mind that if you're just setting a tag that contains the domain name itself - that may be redundant, since we already send the URL including the host.  It's up to you though, perhaps a tag is more meaningful for your situation.

**Comment by mattjohnsonpint** [comment, marked as answer]
@anwarjaved  - Turns out I was wrong about this in my previous replies.  We do have a way to support multiple DSNs in the same application.  There's an example app that does that here:
https://github.com/getsentry/sentry-dotnet/blob/main/samples/Sentry.Samples.Console.Customized/Program.cs

Basically, you need a separate `SentryClient` instance per DSN, then you can apply that client to a particular scope using `SentrySdk.BindClient`.

In your case, you'd have to pre-create and fill a `Dictionary<string, ISentryClient>` to map the domain name to the client instance, then configure scope during the request (or in a custom middleware) to pick the appropriate client and bind to it on that scope.

I haven't tested this myself, but in theory it should work.  Please let us know if you try and how it goes.  Thanks!

**Reply by mattjohnsonpint** [reply]
BTW, I suggest you make your dictionary case-insenstive.

```csharp
var sentryClients = new Dictionary<string, ISentryClient>(StringComparer.OrdinalIgnoreCase);
```

**Comment by anwarjaved** [comment]
@mattjohnsonpint Thank you, Let me try and get back to you if I need further help
